### PR TITLE
More cbook deprecations.

### PR DESCRIPTION
**Pull request by anntzer**
`is_scalar_or_string` was not used anymore, `is_scalar` was only used in one
place (in mplot3d) and I rewrote that snippet into something more robust.

### CODE REVIEW

**Comment by WeatherGod** [comment]
The new logic isn't equivalent, but it is certainly a whole lot easier to understand. I am really not sure what edge cases the old logic was trying to catch here. A tenative +1 from me.

**Comment by anntzer** [comment]
That's why I wrote "more robust".  The old version would get you the incorrect error message when passing incorrect x, y, z with different lengths, for example.

**Comment by dopplershift** [comment]
Jeez, given how much cleaner the new version is, I'm happy so long as no tests break. Don't make it more convoluted until we have tests justifying it.

File: lib/matplotlib/cbook/__init__.py [hunk 1]
```diff
@@ -543,6 +543,7 @@ def file_requires_unicode(x):
         return False
 
 
+@deprecated('2.1')
 def is_scalar(obj):
     """return true if *obj* is not string like and is not iterable"""
     return not is_string_like(obj) and not iterable(obj)
```
Comment (story645) [inline lib/matplotlib/cbook/__init__.py]: is `is_scaler` supposed to be broader than `is_numlike`?

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 2]
```diff
@@ -1503,39 +1503,25 @@ def plot(self, xs, ys, *args, **kwargs):
...
-
-        # First argument is z value
-        elif args and cbook.is_scalar(args[0]):
-            zs = args[argsi]
-            argsi += 1
+        # `zs` can be passed positionally or as keyword; checking with
+        # `_is_string_like` matches the behavior of 2D `plot` (via
+        # `_process_plot_var_args`).
+        if args and not is_string_like(args[0]):
+            zs = args[0]
```
Comment (story645) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: why is `zs = args.pop(0)` a bad idea?

Comment (anntzer) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: b/c there's no tuple.pop?

File: lib/matplotlib/cbook/__init__.py (hunk 1 above)
Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py]: Arbitrary objects (that are not iterable) are scalars, not numlike.  Why?

**Comment by WeatherGod** [comment]
yeah, the results of the tests will be what will convince me that this change is ok.

File: lib/matplotlib/cbook/__init__.py (hunk 1 above)
Comment (story645) [inline lib/matplotlib/cbook/__init__.py]: 'cause that confuses me 'cause aren't scalers numbers? What definition of scalers are we using then?

Comment (anntzer) [inline lib/matplotlib/cbook/__init__.py]: We don't use this function and I exactly don't want to get bogged down into defining what its semantics should be.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 2 above)
Comment (story645) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: :( sorry got mixed up and thought args was a list.

**Comment by WeatherGod** [comment]
ok, the tests pass. Is there a place in the docs where we are documenting deprecations? Or does that happen automatically with the decorator?

**Comment by anntzer** [comment]
I think we just grep for `@deprecated("...")`.

**Comment by WeatherGod** [comment]
Thanks!

**Comment by QuLogic** [comment]
`is_scalar_or_string` is used by default by `cbook.flatten` and that is used in at least `artist.py` and `container.py`. It's causing deprecation warnings when building the docs.

**Pull request merged**